contracts/governance/extensions/GovernorCountingFractional.sol

```diff
@@ -0,0 +1,232 @@
+
+    /**
+     * @dev Record votes with full weight cast for `support`.
+     *
+     * Because this function votes with the delegate's remaining weight, it can only be called once per proposal and
+     * thus does not require any replay protection.
+     */
+    function _countVoteNominal(
+        uint256 proposalId,
+        address account,
+        uint8 support,
+        uint256 weight
+    ) private returns (uint256) {
+        ProposalVote storage details = _proposalVotes[proposalId];
+        details.usedVotes[account] += weight;
+
+        if (support == uint8(GovernorCountingSimple.VoteType.Against)) {
+            details.againstVotes += weight;
+        } else if (support == uint8(GovernorCountingSimple.VoteType.For)) {
+            details.forVotes += weight;
+        } else if (support == uint8(GovernorCountingSimple.VoteType.Abstain)) {
+            details.abstainVotes += weight;
+        }
+
+        return weight;
+    }
+
+    /**
+     * @dev Count votes with fractional weight.
+     *
+     * The `params` argument is expected to be three packed `uint128`:
+     * `abi.encodePacked(uint128(againstVotes), uint128(forVotes), uint128(abstainVotes))`
+     *
+     * This function can be called multiple times for the same account and proposal (i.e. partial/rolling votes are
+     * allowed). For example, an account with total weight of 10 could call this function three times with the
+     * following vote data:
+     *
+     *   * Against: 1, For: 0, Abstain: 2
+     *   * Against: 3, For: 1, Abstain: 0
+     *   * Against: 1, For: 1, Abstain: 1
+     *
+     * Casting votes like this will make the calling account to cast a total of 5 `Against` votes, 2 `For` votes
+     * and 3 `Abstain` votes. Though partial, votes are still final once cast and cannot be changed or overridden.
+     * Subsequent partial votes simply increment existing totals.
+     */
+    function _countVoteFractional(
+        uint256 proposalId,
+        address account,
+        bytes memory params,
+        uint256 weight
+    ) private returns (uint256) {
+        uint128 againstVotes = _extractUint128(params, 0);
+        uint128 forVotes = _extractUint128(params, 1);
+        uint128 abstainVotes = _extractUint128(params, 2);
+
+        uint256 usedWeight = againstVotes + forVotes + abstainVotes;
+        if (usedWeight > weight) {
+            revert GovernorExceedRemainingWeight(account, usedWeight, weight);
+        }
+
+        ProposalVote storage details = _proposalVotes[proposalId];
+        if (againstVotes > 0) {
+            details.againstVotes += againstVotes;
+        }
+        if (forVotes > 0) {
+            details.forVotes += forVotes;
+        }
+        if (abstainVotes > 0) {
+            details.abstainVotes += abstainVotes;
+        }
+        details.usedVotes[account] += usedWeight;
+
+        return usedWeight;
+    }
+
+    function _extractUint128(bytes memory data, uint256 pos) private pure returns (uint128 result) {
```

Comment:
This function concerns me because the length of `data` is ignored and `pos` may be reading out of bounds. The assembly block correctly omits the `memory-safe` annotation, but like I said in a recent issue this globally disables optimizations so it should be avoided.

Is it possible to just check the validity of `pos` against the data length? Sadly I do see it would make the code less efficient by checking the length three times...

Reply:
This is private, and everywhere we call it we first validate the length. Is that not good enough?

Reply:
GIven how frequent it is for people to copy-paste from our code I think we should try to make functions self-contained and well documented even if they're private.

In this case, not only that I think this is the cleanest way of making the assembly block `memory-safe` to avoid disabling optimizations, but also I share the concern with @frangio; although the function is unreachable with invalid parameters, I don't feel comfortable having such a function around.